lib/models/package-info-cache/package-info.js

```diff
@@ -415,6 +421,190 @@ class PackageInfo {
+   This is to be called ONLY on the PackageInfo for a Project.
+
+   @method getBundledProject
+   @return {Set} a Set of {PackageInfo} objects.
+  */
+  getBundledProjectAddons() {
+    if (this.bundledProjectAddons) {
+      return this.bundledProjectAddons;
+    }
+
+    let projectAddons = new Set();
+
+    function findAddons(pkgInfo) {
+      if (projectAddons.has(pkgInfo)) {
+        return;
+      }
+
+      if (!pkgInfo.valid || !pkgInfo.addonMainPath) {
+        return;
+      }
+
+      require(pkgInfo.addonMainPath);
+
+      if (pkgInfo.isBundleRoot) {
+        return;
+      }
+
+      projectAddons.add(pkgInfo);
+
+      let addonPackageList = pkgInfo.discoverAddonAddons();
+      addonPackageList.forEach((pkgInfo) => findAddons(pkgInfo));
+    }
+
+    let addonPackageList = this.discoverProjectAddons();
+    addonPackageList.forEach((pkgInfo) => findAddons(pkgInfo));
+
+    this.bundledProjectAddons = projectAddons;
+
+    return projectAddons;
+  }
+
+  /**
+   * Create an instance of the addon represented by this packageInfo or (if we
+   * are supporting singletons and this is a singleton-enabled addon) check if
+   * we should be creating a proxy instead.
+   * NOTE: we assume that the value of 'isSingleton' does not change between
+   * calls to the constructor! A given addon is either a singleton or not for an
+   * entire run.
+   *
+   * @method getAddonInstance
+   * @param {} parent the addon/project that is to be the direct parent of the
+   * addon instance created here
+   * @param {*} project the project that is to contain this addon instance
+   * @return {Object} the constructed instance of the addon
+   */
+  getAddonInstance(parent, project) {
+    let ctor = this.getAddonConstructor();
+    let addon;
+
+    if (!isSupportingSingletons || !ctor.isSingleton) {
+      // We have a 'regular' addon or we're not supporting singleton addons at all,
+      // so just return a regular instance. It is not in any singleton cache.
+      addon = this.constructAddonInstance(parent, project);
+
+      if (addon.initializeAddons) {
+        addon.initializeAddons();
+      } else {
+        addon.addons = [];
+      }
+
+      return addon;
+    }
+
+    // Find the bundleRoot that will contain this singleton instance.
+    // Find some parent, if there is one, that defines the 'isBundleRoot' property and that
+    // returns a truthy value. 'Project' is always the ultimate bundle.
+
+    // An optimization that lazy engines does that we can do as well. If the bundle
+    // root is not the project, but the project does or is going to have an instance
+    // of this singleton in its own cache, then use the project cache.
+    let bundleRoot = this.findBundleRoot(parent);
+    let bundledProjectAddons = project.getPackageInfo().getBundledProjectAddons();
+
+    if (bundleRoot === project || (bundleRoot !== project && !bundledProjectAddons.has(this))) {
```

Comment:
Let's simplify this conditional a bit (likely by making local variables with good names).

Reply:
For Brenden's second comment - this is the part that is affected by the redesign of what's in PerBundleAddonCache. The intention there is that when a lazy engine is being created, the PBAC will see it's a lazy engine, see if there is already a cache entry (in the cache-by-name I mentioned earlier), and if so, just return a proxy to that instance. The first time a given lazy engine (by name) is requested, it will be created as a real instance and that cache-by-name entry will be created. (So lazy-engine instances behave like other addons do - one "real" copy with real parent, everybody else a Proxy).

Reply:
I'm only keeping this open now because I need to add the check in the caching code for whether deduping is enabled or not. I've addressed the rest of it.

Reply:
I've now added the check in PerBundleAddonCache's function bundleRootOwnsInstance(). I just have a const that equals `process.env.EMBER_ENGINES_ADDON_DEDUPE`. Since it's in process I don't need to do any trickery with checking if we can require `ember-engines` (though that would ultimately be preferable, if ember-engines exported this process var, so we didn't have to rely on how it's named.